OP-1332 OP-1199 OP-1200 Pharmacy Inventory feature

mwithi opened the pull request:
See OP-1199 and OP-1200 (also addresses OP-1332).

Paired with https://github.com/informatici/openhospital-core/pull/1480.

Todo:
- [x] To always show at least "open" (draft, validated) inventories regardless the date range, when the filter is set to "All"
- [x] To improve "Add Medicals" dialog's layout
- [x] Add column to show real - theoretical difference (if any)

mwithi committed:
Merge pull request #2086 from informatici/develop

Merge with develop

JantBogard committed:
Merge branch 'OP-1200-Wards-Inventory' into OP-1193_OP-1194_OP-1197

JantBogard committed:
Applying suggestion for change

JantBogard committed:
Applying suggestion to change

JantBogard committed:
Applying suggestion to change

JantBogard committed:
Applying suggestion to change

mwithi committed:
Merge pull request #2080 from JantBogard/OP-1187

OP-1187 | Printing Inventory

mwithi committed:
Merge branch 'develop' into OP-1199-Main-Store-Inventory

mwithi committed:
Merge branch 'develop' into OP-1200-Wards-Inventory

mwithi committed:
Fix

mwithi committed:
Merge branch 'OP-1200-Wards-Inventory' into OP-1193_OP-1194_OP-1197

JantBogard committed:
Applying suggestion to change

JantBogard committed:
Merge branch 'OP-1193_OP-1194_OP-1197' of https://github.com/JantBogard/openhospital-gui into OP-1193_OP-1194_OP-1197

ArnaudFofou committed:
OP-1305 | Confirm Inventory (#2070)

* add confirm inventory feature

* update bundle/language_en.properties

* revert changes

* format the code

* Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* update button order

* apply suggestion

* Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* remove unused imports

* add possibilty to update prise of existing lot

* Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* add check of existing reference on stock movement

* add another condition to enable lot button

* fix error on bundler

* Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* add check of reference number on validateInventory

* reset param of settings.properties.dist

* update InventoryEdit

* revert code

* remove non used import

* update inventory during validation

* fix bugs

* fix bugs on InventoryBrowser

* add TimeTools to date properties

* add missng bundle key

* apply suggestion

* update InventoryEdit

* revert settings.properties.dist

* update InventoryEdit

* update  src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* updat InventoryEdit.java

* add method the know if all medical are in inventory

* add logic to select and deselect specific and all radio button

* update the logic when saving inventory

* update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* Update bundle/language_en.properties

* fix bug during save action

* fix bug

* charge Iterator by ListIterator

* Update bundle/language_en.properties

* Update src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java

* Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* remove empty comment

* fix conflit after merge

---------

Co-authored-by: ArnaudFonzam <gui.fofou@umi2growcameroun.com>
Co-authored-by: FOFOU FONZAM Gui Arnaud <101590821+ArnaudFonzam@users.noreply.github.com>
Co-authored-by: Alessandro Domanico <alessandro.domanico@yahoo.it>
Co-authored-by: David B Malkovsky <david@malkovsky.org>

JantBogard committed:
Add new message bundle

JantBogard committed:
Merge branch 'develop' into OP-1193_OP-1194_OP-1197

JantBogard committed:
Fix indentations issue

JantBogard committed:
Fix bug occur when i update CORE

JantBogard committed:
Merge branch 'OP-1193_OP-1194_OP-1197' of https://github.com/JantBogard/openhospital-gui into OP-1193_OP-1194_OP-1197

Commit:
Enhance product selections in edit inventory

Commit:
Merge branch 'develop' into OP-1304

ArnaudFofou committed:
Merge branch 'informatici:develop' into OP-1304

ArnaudFofou committed:
Update InventoryEdit.java

ArnaudFofou committed:
Update InventoryEdit.java

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Commit:
Merge branch 'OP-1304' of https://github.com/ArnaudFofou/openhospital-gui into OP-1304

JantBogard committed:
Applying suggestion for change

JantBogard committed:
Applying suggestion for change

Commit:
update InventoryEdit

JantBogard committed:
Applying suggestion for change

JantBogard committed:
Applying suggestion for change

JantBogard committed:
Applying suggestion for change

JantBogard committed:
Revert "Applying suggestion for change"

JantBogard committed:
Fix format issues

JantBogard committed:
Applying suggestion to change

JantBogard committed:
formating

JantBogard committed:
formating

JantBogard committed:
formating

JantBogard committed:
formating

JantBogard committed:
formating

JantBogard committed:
formatting

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: David B Malkovsky <david@malkovsky.org>

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: David B Malkovsky <david@malkovsky.org>

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: David B Malkovsky <david@malkovsky.org>

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: David B Malkovsky <david@malkovsky.org>

Commit:
apply suggestion

Commit:
Merge branch 'OP-1304' of https://github.com/ArnaudFofou/openhospital-gui into OP-1304

JantBogard committed:
Fix format issues

JantBogard committed:
Merge branch 'develop' into OP-1193_OP-1194_OP-1197

Commit:
apply suggestion

JantBogard committed:
Fix format issues

Commit:
apply suggestion

Commit:
enhance strategie to fetch medicals

JantBogard committed:
Fix format issues

mwithi committed:
Merge branch 'develop' into OP-1199-Main-Store-Inventory

JantBogard committed:
Remove all necessary field

JantBogard committed:
Merge branch 'develop' into OP-1193_OP-1194_OP-1197

JantBogard committed:
Add possibility to actualize inventory row when we validate inventory and fix a few bugs

Commit:
apply suggestion

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: Alessandro Domanico <alessandro.domanico@yahoo.it>

ArnaudFofou committed:
OP-1368 | Make Inventory status language independent (#2098)

* update status management

* update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* apply suggestion

* apply suggestion

* update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

* align status on alphabetical order

---------

Co-authored-by: ArnaudFofou <gui.fofou@umi2growcameroun.com>

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: David B Malkovsky <david@malkovsky.org>

Commit:
apply suggestion

Commit:
Merge branch 'develop' into OP-1304

Commit:
fix conflict

Commit:
apply suggestion

Commit:
apply suggestion

JantBogard committed:
Applying a few change

Commit:
fix suggestion

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Commit:
enhance the logic of Select Button

mwithi committed:
Merge branch 'develop' into OP-1200-Wards-Inventory

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: Alessandro Domanico <alessandro.domanico@yahoo.it>

ArnaudFofou committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Co-authored-by: Alessandro Domanico <alessandro.domanico@yahoo.it>

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java'

Commit:
remove unused import

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java'

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java'

Commit:
update InventoryEdit

JantBogard committed:
Add possibility to get all medical in chooese medical and add manage lot

mwithi committed:
Fix formatting

mwithi committed:
Remove redundant code and translate French comments

mwithi committed:
Remove redundant code

mwithi committed:
Fix

mwithi committed:
Merge branch 'OP-1304' into OP-1304

mwithi committed:
Update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

mwithi committed:
Merge pull request #2091 from ArnaudFofou/OP-1304

OP-1304 | Enhance product selections in edit inventory

mwithi committed:
Merge pull request #2106 from mwithi/OP-1304

OP-1304 Small changes

mwithi committed:
Improve cell rendering

mwithi committed:
Fix typo

mwithi committed:
Improve cell rendering in browser

mwithi committed:
Change columns headers

mwithi committed:
Refactor variables

mwithi committed:
Standardize bundles

Commit:
enhance product selection

ArnaudFofou committed:
Merge branch 'OP-1199-Main-Store-Inventory' into OP-1304

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Commit:
Merge branch 'OP-1304' of https://github.com/ArnaudFofou/openhospital-gui into OP-1304

Commit:
enhance code and avoid duplicated instruction

JantBogard committed:
Fixed bugs

JantBogard committed:
Fixed bugs

mwithi committed:
Add missing bundle and make draft inventories more evident

mwithi committed:
Fix pagination

mwithi committed:
Restore PAGE_SIZE

JantBogard committed:
Fixed bugs

mwithi committed:
Fix formatting

mwithi committed:
Merge pull request #2109 from ArnaudFofou/OP-1304

OP-1304 | Enhance the Code with also to enhance Product Selection

mwithi committed:
Merge branch 'OP-1199-Main-Store-Inventory' into
mwithi_OP-1199_small_improvements

JantBogard committed:
Fixed null pointer exception when you not exit code in specific product textField

mwithi committed:
Merge branch 'develop' into OP-1199-Main-Store-Inventory

JantBogard committed:
bloc the possibility to validate inventory with new lot having real quantity

Commit:
fix bug and enhance the code

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

ArnaudFofou committed:
Update bundle/language_en.properties

Co-authored-by: David B Malkovsky <david@malkovsky.org>

Commit:
change info  to error  on messageDialog

Commit:
Merge branch 'OP-1376' of https://github.com/ArnaudFofou/openhospital-gui into OP-1376

mwithi committed:
Merge pull request #2116 from informatici/develop

Align with develop

mwithi committed:
Merge branch 'OP-1199-Main-Store-Inventory' into mwithi_OP-1199_small_improvements

mwithi committed:
Initialization to null is redundant

mwithi committed:
The result of startIndex / PAGE_SIZE is always zero (startIndex is 0).

So the result is always 1

mwithi committed:
use a formatted message

Commit:
update the message bundle

Commit:
enhance the code

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

JantBogard committed:
Fix unpredicted inventoryEdit close behavior

Commit:
update checkParam method

JantBogard committed:
Remove "do you want to continue editing" message when you save an inventory.

Commit:
update  src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Commit:
Merge branch 'develop' into OP-1376

Commit:
fix bugs and add message to avoid validate or confirm inventory without products'

Commit:
update bundle/language_en.properties

Commit:
Merge branch 'develop' into OP-1376

Commit:
update src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java

Commit:
disable codeTextField and selectButton on edit mode

Commit:
check if something has been update after action confirm'

Commit:
update the date of the Licence

Commit:
update the date of the  Licence

Commit:
update the status if the table is not actualise during validation

mwithi committed:
Merge pull request #2115 from ArnaudFofou/OP-1376

OP-1376 |  Fix some Bug

mwithi committed:
Merge branch 'OP-1199-Main-Store-Inventory' into OP-1199_small_improvements

mwithi committed:
Merge pull request #2085 from JantBogard/OP-1193_OP-1194_OP-1197

OP-1193 | OP-1194 | OP-1197 | OP-1198 | Update, Delete and validate Ward inventory

mwithi committed:
Merge pull request #2107 from mwithi/OP-1199_small_improvements

OP-1199 Small improvements

mwithi committed:
Fix total cost calculation as it is BigDecimal now

Commit:
add button confirm and implement the logic of that button

Commit:
update message bundle and the logic for confirmButton

ArnaudFofou committed:
Update bundle/language_en.properties

Co-authored-by: David B Malkovsky <david@malkovsky.org>

ArnaudFofou committed:
Update bundle/language_en.properties

Co-authored-by: David B Malkovsky <david@malkovsky.org>

ArnaudFofou committed:
Update bundle/language_en.properties

Co-authored-by: David B Malkovsky <david@malkovsky.org>

Commit:
update  bundle/language_en.properties

Commit:
Merge branch 'OP-1376' of https://github.com/ArnaudFofou/openhospital-gui into OP-1376

Commit:
update the licence date

mwithi committed:
Merge pull request #2120 from informatici/develop

OP-1306 Fix alcool -> alcohol (#2117)

JantBogard committed:
OP-1195 | Print ward Inventory (#2119)

* Print inventory ward report

* fix bug

Commit:
add possibility to choose lot during inventory

Commit:
Merge branch 'develop' into OP-1376

Commit:
check the status of existing war inventory before give possibility to create new one

Commit:
Merge remote-tracking branch 'upstream/OP-1200-Wards-Inventory' into OP-1376

ArnaudFofou committed:
Merge branch 'OP-1200-Wards-Inventory' into OP-1376

Commit:
Merge branch 'OP-1376' of https://github.com/ArnaudFofou/openhospital-gui into OP-1376

Commit:
fix buggs

Commit:
fix null pointer exception

Commit:
change lot.getMainStoreQuantity() to lot.getWardsTotalQuantity()

Commit:
fix bug

Commit:
remove column qty and not on the lot table

mwithi committed:
Merge pull request #2118 from ArnaudFofou/OP-1376

OP-1374 | Confirm ward Inventory

mwithi committed:
Apply suggestions from code review

Co-authored-by: David B Malkovsky <david@malkovsky.org>

mwithi committed:
Merge branch 'develop' into OP-1199-Main-Store-Inventory

mwithi committed:
Fix default JLabel

mwithi committed:
Add blank line

mwithi committed:
Remove unused bundle

mwithi committed:
Apply suggestions from code review

Co-authored-by: David B Malkovsky <david@malkovsky.org>

mwithi committed:
Fix old fashion code

mwithi committed:
Add @Override annotation

mwithi committed:
Use interface constant

mwithi committed:
Use lambda

mwithi committed:
Remove redundant 'null', 'public' and 'return'

mwithi committed:
Add @Override

mwithi committed:
Remove unnecessary casting

mwithi committed:
Apply code review suggestions

mwithi committed:
Add @Override

mwithi committed:
Apply code review suggestions

mwithi committed:
Apply suggestions from code review

Co-authored-by: David B Malkovsky <david@malkovsky.org>

mwithi committed:
Apply suggestions from code review

mwithi committed:
Merge pull request #2125 from informatici/OP-1199-Main-Store-Inventory

OP-1199 Main store inventory

mwithi committed:
Merge branch 'OP-1332_OP-1199_OP-1200' into OP-1200-Wards-Inventory

mwithi committed:
Merge pull request #2126 from informatici/OP-1200-Wards-Inventory

OP-1200 wards inventory

mwithi committed:
Reducing differences for future refactoring

mwithi commented:
I'm not very sure that this appeareance is the best we can achieve though...

@dbmalkovsky @ArnaudFonzam do you think could be a good idea to just move the new two buttons inside the other three above? As normal buttons, they could be parametrized and so call a single "Browser" which differs only in the type of inventory (`InventoryType`) and few other things...

![image](https://github.com/user-attachments/assets/6ad7700f-6646-4c28-8007-9ae6466b450e)

mwithi committed:
Use same color formatter

mwithi committed:
Change title

mwithi committed:
Sort bundles

mwithi committed:
Remove duplicates and fix some bundles

mwithi committed:
Fix bundles

mwithi committed:
Fix bundles

mwithi committed:
Fix bundles

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -195,6 +195,9 @@ angal.agetype.somerangesarenotdefinedpleasecheckthevalues.msg
 angal.anamnesis.alcohol.no.txt                                                                         = No
 angal.anamnesis.alcohol.txt                                                                            = Alcohol
 angal.anamnesis.alcohol.yes.txt                                                                        = Yes
+angal.anamnesis.alvo.normal.txt                                                                        = Normal
```
line 175: `transfered` -> `transferred`

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -195,6 +195,9 @@ angal.agetype.somerangesarenotdefinedpleasecheckthevalues.msg
 angal.anamnesis.alcohol.yes.txt                                                                        = Yes
+angal.anamnesis.alvo.normal.txt                                                                        = Normal
+angal.anamnesis.alvo.other.txt                                                                         = Other
+angal.anamnesis.alvo.txt                                                                               = Alvo
```
line 209:  `Cardiovascolaris` -> `Cardiovascular`

dbmalkovsky replied:
line 215: `Ortoreumatologic` (Swedish) -> `Orthorheumatology` (English)

mwithi committed:
Remove orphan keys

mwithi committed:
Fix more bundles

ArnaudFofou commented on src/main/java/org/isf/medicalinventory/gui/InventoryWardBrowser.java:
```diff
@@ -0,0 +1,683 @@
+	private JButton next;
+	private JButton previous;
+	private JComboBox<Integer> pagesComboBox = new JComboBox<>();
+	private JLabel ofPagesLabel = new JLabel(MessageBundle.formatMessage("angal.common.pages.fmt.txt"), 1);
```
```suggestion
	private JLabel ofPagesLabel = new JLabel(MessageBundle.formatMessage("angal.common.pages.fmt.txt", 1));
```

ArnaudFofou commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,670 @@
+	private JButton next;
+	private JButton previous;
+	private JComboBox<Integer> pagesComboBox = new JComboBox<>();
+	private JLabel ofPagesLabel = new JLabel(MessageBundle.formatMessage("angal.common.pages.fmt.txt"), 1);
```
```suggestion
	private JLabel ofPagesLabel = new JLabel(MessageBundle.formatMessage("angal.common.pages.fmt.txt", 1));
```

mwithi committed:
Apply suggestions from code review

Co-authored-by: FOFOU FONZAM Gui Arnaud <gui.fofou@uni2growcameroun.com>

ArnaudFofou commented:
> I'm not very sure that this appeareance is the best we can achieve though...
> 
> @dbmalkovsky @ArnaudFonzam do you think could be a good idea to just move the new two buttons inside the other three above? As normal buttons, they could be parametrized and so call a single "Browser" which differs only in the type of inventory (`InventoryType`) and few other things...
> 
> ![image](https://private-user-images.githubusercontent.com/2938553/407664444-6ad7700f-6646-4c28-8007-9ae6466b450e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxNjcxNjAsIm5iZiI6MTczODE2Njg2MCwicGF0aCI6Ii8yOTM4NTUzLzQwNzY2NDQ0NC02YWQ3NzAwZi02NjQ2LTRjMjgtODAwNy05YWU2NDY2YjQ1MGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjlUMTYwNzQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTJhYTgxMWE1YzRlZDgzZTVkMDBkNzMyMGRjNjYyNGRiMDI4MWIwODY3MWFiYjkxMWY2NGI4MjUzODc4MmNkOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.d0ND1mQy_-GPSmNqBe5LYKKdZholbQKtPAhdW5XbkYk)

For me if we move those two buttons inside the other three it is not good we will have many intermediaire window when user just need to do and inventory.

please Have you see that ward Inventory button don't have shortcut. may be the problem is that the Shortcut used is already use on this window. I think we should update the menu item row related to ward inventory button to set the non used shortcut. WDYT?

mwithi committed:
Add missing bundles

mwithi committed:
Remove orphan bundles

mwithi commented:
> please Have you see that ward Inventory button don't have shortcut. may be the problem is that the Shortcut used is already use on this window. I think we should update the menu item row related to ward inventory button to set the non used shortcut. WDYT?

Done, and also inverted item order, see https://github.com/informatici/openhospital-core/pull/1474. Thanks!

mwithi commented:
> For me if we move those two buttons inside the other three it is not good we will have many intermediaire window when user just need to do and inventory.

True, but in this way we have a lot of code duplication so when you want to change something you have to do it in two places. On the other hand, you have that the inventory is not a so frequent operation (generally yearly, hopefully quarterly, rarely monthly).
Anyway also other solution to avoid code duplication are warmly welcomed.

mwithi committed:
Make columns identical and remove unnecessary bundles

mwithi committed:
Reduce differences and bundles

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,124 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventory.cannotsaveinventorywithoutproducts.msg                                                 = Cannot save inventory without products.
+angal.inventory.cannotvalidateinventorywithoutproducts.msg                                             = Cannot validate inventory without products.
+angal.inventory.chargetype.label                                                                       = Charge Type
+angal.inventory.chargetype.label                                                                       = Charge Type:
```
Remove this line; it duplicates line 695

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,124 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventory.deletion.success.msg                                                                   = The inventory has been successfully marked as deleted.
+angal.inventory.destination.label                                                                      = Destination
+angal.inventory.dischargetype.label                                                                    = Discharge Type
+angal.inventory.dischargetype.label                                                                    = Discharge Type:
```
Remove this line; it duplicates line 707

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,124 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventory.dischargetype.label                                                                    = Discharge Type
+angal.inventory.dischargetype.label                                                                    = Discharge Type:
+angal.inventory.donenoteditable.msg                                                                    = A confirmed inventory can not be modified. 
+angal.inventory.donenoteditable.msg                                                                    = The inventory with status done is not editable.
```
line 709 and 710 have the same key but different strings.

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,124 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventory.pleaseselectinventory.msg                                                              = Please select inventory.
+angal.inventory.pleaseselectonlyoneinventory.msg                                                       = Please select only one inventory.
+angal.inventory.pleasevalidateinventoryagainsbeforeconfirmation.msg                                    = Please validate the inventory again before confirming it.
+angal.inventory.pleasevalidateinventoryagainsbeforeconfirmation.msg                                    = Please validate this inventory again before confirmation.
```
lines 748 and 749 have the same key but slightly different text strings.

mwithi committed:
Apply code review suggestions

mwithi committed:
Reduce differences

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,118 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventoryrow.lotinformation.title                                                                = Lot Information
+angal.inventoryrow.medicaltype.txt                                                                     = Medical Type
+angal.inventoryrow.pleaseselectatleastoneinventoryrow.msg                                              = Please select at least one inventory row.
+angal.inventoryrow.pleaseselectatleastoneinventoryrow.msg                                              = Please select at least one inventory row.
```
Please delete one of these two rows (787 and 788); the key and the text is the same.

mwithi committed:
Fix bundles (revert unwanted deletions)

mwithi committed:
Remove duplicated bundles

mwithi committed:
Reduce diff

mwithi committed:
Remove useless method

mwithi committed:
Reduce diff

mwithi committed:
Fix status filter

mwithi committed:
Reduce diff and rename fields and methods

mwithi committed:
Reduce diff

mwithi committed:
Fix bug when adding all products

mwithi committed:
There's no need to check the status before to set it anyway

mwithi committed:
Change status only upon success

ArnaudFofou commented on src/main/java/org/isf/medicalinventory/gui/InventoryWardEdit.java:
```diff
@@ -0,0 +1,1821 @@
+						MessageDialog.error(null, "angal.inventory.referencealreadyused.msg");
+						return;
+					}
+					inventory = new MedicalInventory();
```
```suggestion
					inventory = new MedicalInventory();
					wardSelected = (Ward)wardComboBox.getSelectedItem();
					inventory.setWard(wardSelected.getCode());
```

mwithi replied:
the `setWard` method is already on the line 498, `wardSelected` is global and it should be already set at this point because ward selection is compulsory before to even start a new inventory, so in theory we could also remove the `null` check...

Have you experienced a bug? 
Anyway we have also a global `wardId`, which seems just unnecessary if we already have `wardSelected`... I guess we can simplify a lot the logic here...

ArnaudFofou replied:
yes I have the error after the inventory was created  after checking I see that ward column was null even as we set the value before create its. I think the problem is because the value of selected ward is not set to the variable wardSelect and the value is null during creation. look the listener of wardComboBox you didn't set the value of the wardSelected but you used wardId, so you can use  wardSelected  instead of wardId on wardCombox Listener to avoid a lot of variable doing the same things.

mwithi replied:
As for now I've published a quick fix, please try.

mwithi replied:
hello @ArnaudFofou did you try if the latest code on this branch solved your error?
Please update, just added some changes to ackknowledge the new develop...

mwithi committed:
Fix ward selection

mwithi committed:
Align with latest develop

mwithi committed:
Align with latest develop

mwithi committed:
Merge branch 'develop' into OP-1332_OP-1199_OP-1200

mwithi committed:
Make new inventories creation transactional

mwithi committed:
Merge branch 'OP-1332_OP-1199_OP-1200' of https://github.com/informatici/openhospital-gui.git into OP-1332_OP-1199_OP-1200

mwithi committed:
Fixed and simplified lot management, save and update

also:
- done some refactoring
- no need to check reference here since it's done in the validate step
- fixed one bundle

mwithi committed:
Fix for new lots loading and add check for reference length

ArnaudFofou committed:
OP-1385 | Fix error on inventory report (#2130)

* update inventory and inventoryward report and add missing inventory bundle

* remove bundle and update the inventory report

* update inventoryWard properties

* update ward inventory report

---------

Co-authored-by: David B Malkovsky <david@malkovsky.org>
Co-authored-by: ArnaudFofou <gui.fofou@umi2growcameroun.com>

mwithi committed:
Indent one comment

mwithi committed:
Limit reference JTextfield

mwithi committed:
Remove useless bundle

mwithi committed:
Fix inventory actualization

mwithi committed:
Fix date update after actualization

mwithi committed:
Level out labels and translations

mwithi committed:
Main inventory enhancements.

Including:
- Inventory date and time save and update
- Enable user to perform partial or broad validation
- Remove unnecessary questions
- Detect inventory date changes
- Done some refactoring

mwithi committed:
Fix bundle and remove unnecessary question

mwithi committed:
Enhancement for Ward Inventory.

Including:
- As much as similar to Main Inventory (TODO: can be merged?)
- Fix some typos
- Improved TimeTools comparison
- Fix some bundles
- Also fix the way WardPharmacyNew save movements dates

mwithi commented:
Guys, I did my best to make it at least functioning and consistent among Main and Ward scenarios.
Some code can be for sure optimized/simplified though... I appreciate code review and some hard test (with most weird cases of concurrent changes on the stocks!)
Next change will be a way to always show at least "open" inventories despite the date range, when the filter is set to "All".
Thanks in advance!

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -1727,9 +1869,9 @@ angal.userbrowser.resetpassword.btn.key
 angal.userbrowser.selectgroup.label                                                                    = Select group:
-angal.userbrowser.alreadysoftdeleted.msg                                                               = The user is already set as deleted.
-angal.userbrowser.step1.pleaseinsertanew.password.msg                                                  = Please insert a new password.
+angal.userbrowser.softdeleteuser.fmt.msg                                                               = Could not delete the user '{0}' because it is used somewhere. Do you want to mark it as deleted ?
```
```suggestion
angal.userbrowser.softdeleteuser.fmt.msg                                                               = Could not delete the user '{0}' because it is used somewhere. Do you want to mark it as deleted?
```

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -1322,10 +1464,10 @@ angal.operationrow.not.found.msg
 angal.operationrowedit.updateerror                                                                     = Error occurs during processing
-angal.operationrowedit.updatesucces                                                                    = Operation updated
+angal.operationrowedit.updatesuccess                                                                   = Operation updated.
 angal.operationrowedit.warningdateafter                                                                = You must choose a date after OPD'S date!
```
```suggestion
angal.operationrowedit.warningdateafter                                                                = You must choose a date after OPD's date.
```

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -645,7 +655,7 @@ angal.groupsbrowser.mainmenu.txt
 angal.groupsbrowser.newgroup.title                                                                     = New Group
 angal.groupsbrowser.pleaseinsertavalidusergroupname.msg                                                = Please insert a valid user group name.
-angal.groupsbrowser.alreadysoftdeleted.msg                                                             = The group is already set as deleted.
+angal.groupsbrowser.softdeletegroup.fmt.msg                                                            = Could not delete the group '{0}' because it is used somewhere. Do you want to mark it as deleted ?
```
```suggestion
angal.groupsbrowser.softdeletegroup.fmt.msg                                                            = Could not delete the group '{0}' because it is used somewhere. Do you want to mark it as deleted?
```

mwithi committed:
Apply suggestions from code review

Co-authored-by: David B Malkovsky <david@malkovsky.org>

dbmalkovsky commented on src/main/java/org/isf/medicalstockward/gui/WardPharmacyNew.java:
```diff
@@ -319,16 +321,16 @@ public boolean isCellEditable(int arg0, int arg1) {
 		double wardQty = medWard.getQty();
 		if (qty > wardQty) {
-			MessageDialog.error(this, "angal.medicalstock.movementquantityisgreaterthanthequantityof.msg");
+			MessageDialog.error(this, "angal.medicalstock.movementquantityisgreaterthanthequantityof.fmt.msg");
```
is this is a `fmt` type message, there should be some parameter/variable also to substitute into the message shouldn't there?

mwithi replied:
good catch! yes, two parameteres... solved, thanks!

mwithi committed:
Fix fmt bundle

mwithi committed:
Fix reference field in "view" mode

mwithi committed:
Make status to 'draft' when validation fails upon confirmation.

Also fix validation just after a save operation.

mwithi committed:
Fix one bundle

mwithi committed:
Clean bundles and rename "inventoryrow" key

mwithi committed:
Fix "Select Medical" dialog's layout

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+			int eventID = itemEvent.getStateChange();
+
+			if (eventID == ItemEvent.SELECTED) {
+				int page_number = (Integer) pagesComboBox.getSelectedItem();
```
```suggestion
				int pageNumber = (Integer) pagesComboBox.getSelectedItem();
```
local variables should use camel case and not underscores

mwithi replied:
You're right, thanks!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+					previous.setEnabled(false);
+				} else {
+					previous.setEnabled(true);
+				}
```
same type of comment as for line 176

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+					next.setEnabled(false);
+				} else {
+					next.setEnabled(true);
+				}
```
same type of comment as for line 176

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+					next.setEnabled(false);
+				} else {
+					next.setEnabled(true);
+				}
```
same type of comment as for line 176

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+	private JPanel panelHeader;
+	private JPanel panelFooter;
+	private JPanel panelContent;
+	private JButton jButtonClose;
```
make this a local variable in the method `getCloneButton()`

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+	private JPanel panelFooter;
+	private JPanel panelContent;
+	private JButton jButtonClose;
+	private JButton jButtonNew;
```
make this a local variable in the method `getNewButton()`

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+					next.setEnabled(false);
+				} else {
+					next.setEnabled(true);
+				}
```
why not replace lines 172-176 with:

<pre>
next.setEnabled(startIndex + PAGE_SIZE > totalRows ? false : true);
</pre>

mwithi replied:
because is much less readable :)

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+			}
+			if (draftMedicalInventories.isEmpty() && validatedMedicalInventories.isEmpty()) {
+				InventoryEdit inventoryEdit = new InventoryEdit();
+				InventoryEdit.addInventoryListener(InventoryBrowser.this);
```
```suggestion
				InventoryEdit.addInventoryListener(this);
```
unnecessary qualifier in this context

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+			if (draftMedicalInventories.isEmpty() && validatedMedicalInventories.isEmpty()) {
+				InventoryEdit inventoryEdit = new InventoryEdit();
+				InventoryEdit.addInventoryListener(InventoryBrowser.this);
+				inventoryEdit.showAsModal(InventoryBrowser.this);
```
```suggestion
				inventoryEdit.showAsModal(this);
```
unnecessary qualifier in this context

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+				return;
+			}
+			InventoryEdit inventoryEdit = new InventoryEdit(inventory, "update");
+			InventoryEdit.addInventoryListener(InventoryBrowser.this);
```
```suggestion
			InventoryEdit.addInventoryListener(this);
```
unnecessary qualifier in this context

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+			}
+			InventoryEdit inventoryEdit = new InventoryEdit(inventory, "update");
+			InventoryEdit.addInventoryListener(InventoryBrowser.this);
+			inventoryEdit.showAsModal(InventoryBrowser.this);
```
```suggestion
			inventoryEdit.showAsModal(this);
```
unnecessary qualifier in this context

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+			if (selectedRow > -1) {
+				inventory = inventoryList.get(selectedRow);
+				InventoryEdit inventoryEdit = new InventoryEdit(inventory, "view");
+				InventoryEdit.addInventoryListener(InventoryBrowser.this);
```
```suggestion
				InventoryEdit.addInventoryListener(this);
```
unnecessary qualifier in this context

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+				inventory = inventoryList.get(selectedRow);
+				InventoryEdit inventoryEdit = new InventoryEdit(inventory, "view");
+				InventoryEdit.addInventoryListener(InventoryBrowser.this);
+				inventoryEdit.showAsModal(InventoryBrowser.this);
```
```suggestion
				inventoryEdit.showAsModal(this);
```
unnecessary qualifier in this context

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+		jButtonEdit.setMnemonic(MessageBundle.getMnemonic("angal.common.edit.btn.key"));
+		jButtonEdit.setEnabled(false);
+		jButtonEdit.addActionListener(actionEvent -> {
+			MedicalInventory inventory = new MedicalInventory();
```
initialization is redundant as it is set in line 379; why not just move the definition of the variable to that line?

mwithi replied:
Done, thanks!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+		jButtonView.setMnemonic(MessageBundle.getMnemonic("angal.inventory.view.btn.key"));
+		jButtonView.setEnabled(false);
+		jButtonView.addActionListener(actionEvent -> {
+			MedicalInventory inventory = new MedicalInventory();
```
similar comment to line 369

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+				inventoryEdit.showAsModal(InventoryBrowser.this);
+			} else {
+				MessageDialog.error(null, "angal.inventory.cannotcreateanotherinventorywithstatusdraft.msg");
+				return;
```
this is a `void` method so a `return` is not needed

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+		lotsSaved.clear();
+	}
+
+	private boolean checkParametersChanges(String wardCode, String chargeCode, String dischargeCode, Integer supplierId, String reference, LocalDateTime date) {
```
the parameter `String reference` is not used in the method; it should be removed

mwithi replied:
it should... fixed. Thanks!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+
+	private String askAllMedicalsOrList(String question) {
+		String option = (String) MessageDialog.inputDialog(null, null, allMedicalsOrList, null, question);
+		return option;
```
there is no need to assign to a local variable; just immediately return the result of the expression

mwithi replied:
Done, thanks!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+		for (MedicalInventoryRow row : inventoryRowSearchList) {
+			inventorySet.add(row.getMedical());
+		}
+		return medicals.size() == inventorySet.size() ? true : false;
```
remove the unnecessary and redundant boolean literals; the comparison itself returns true and false.

mwithi replied:
Thanks!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+				// Add ActionListener to each button
+				sortedActionMap.forEach((key, value) -> {
+					rightPanel.add(key);
+				});
```
useless curly braces
<pre>
sortedActionMap.forEach((key, value) -> rightPanel.add(key););
</pre>

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			|| (destination != null && isMismatch(destination.getCode(), wardCode))
+			|| (chargeType != null && isMismatch(chargeType.getCode(), chargeCode))
+			|| (dischargeType != null && isMismatch(dischargeType.getCode(), dischargeCode))
+			|| (supplier != null && supplier.getSupId() != supplierId)
```
boxed types should use `equals()` and not `==` because it is not comparing actual values by locations in memory.

mwithi replied:
Thanks!!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+	private JLabel addMedicalsByLabel;
+	private JLabel destinationLabel;
+	private JTextField referenceTextField;
+	private JTextField jTextFieldEditor;
```
Move the variable  into  the method `getJTableInventoryRow()` as that is the only place it is used

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			}
+			chargeCombo.addActionListener(actionEvent -> {
+				chargeType = (MovementType) chargeCombo.getSelectedItem();
+			});
```
remove useless curly braces around statement

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+				supplierCombo.addItem(null);
+				for (Supplier supplier : suppliers) {
+					supplierCombo.addItem(supplier);
+					if (inventory != null && supplier.getSupId() == inventory.getSupplier()) {
```
boxed types should use `equals()` and not `==` because it is not comparing actual values by locations in memory.

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			}
+			dischargeCombo.addActionListener(actionEvent -> {
+				dischargeType = (MovementType) dischargeCombo.getSelectedItem();
+			});
```
remove useless curly braces around statement

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			}
+			supplierCombo.addActionListener(actionEvent -> {
+				supplier = (Supplier) supplierCombo.getSelectedItem();
+			});
```
remove uselss curly braces around statement

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+	private JComboBox<Ward> getJComboDestination() {
+		Ward destinationSelected = null;
+		if (destinationCombo == null) {
+			destinationCombo = new JComboBox<Ward>();
```
```suggestion
			destinationCombo = new JComboBox<>();
```
replace type specification with `<>`

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			}
+			destinationCombo.addActionListener(actionEvent -> {
+				destination = (Ward) destinationCombo.getSelectedItem();
+			});
```
remove useless curly braces around statement

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			medicalTypeSelected = (MedicalType) medicalTypeComboBox.getSelectedItem();
+			medicalTypeComboBox.addActionListener(actionEvent -> {
+				medicalTypeSelected = (MedicalType) medicalTypeComboBox.getSelectedItem();
+			});
```
remove useless curly braces around statement

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+			}
+			jCalendarInventoryDate.addDateTimeChangeListener(event -> {
+				dateInventory = jCalendarInventoryDate.getLocalDateTime();
+			});
```
remove useless curly braces around statement

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:
```diff
@@ -0,0 +1,2201 @@
+
+	public interface InventoryListener extends EventListener {
+
+		void InventoryInserted(AWTEvent e);
```
don't method names start with lower case characters?

mwithi replied:
absolutely! Thanks!

dbmalkovsky commented on src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:
```diff
@@ -0,0 +1,671 @@
+	private JComboBox<Integer> pagesComboBox = new JComboBox<>();
+	private JLabel ofPagesLabel = new JLabel(MessageBundle.formatMessage("angal.common.pages.fmt.txt", 1));
+	private static int PAGE_SIZE = 24;
+	private int startIndex = 0;
```
```suggestion
	private int startIndex;
```
redundant field initialization

mwithi committed:
Apply code review suggestions

mwithi committed:
Rename three methods

mwithi committed:
Rename two fields

mwithi committed:
Fix one bundle

mwithi committed:
Show both draft and validated inventories in BLUE color

mwithi committed:
Add DIFFERENCE column

mwithi committed:
Improve ward inventory selection and check

Also renamed one inventory field String ward -> String wardCode

mwithi committed:
Return expression directly instead of using a variable

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,116 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventory.addamedical.label                                                                      = Add a medical
+angal.inventory.addmedicalsby.label                                                                    = Add medicals by
+angal.inventory.allinventoryrowshouldhavelotbeforevalidation.msg                                       = All the inventory rows should have lots before validation.
+angal.inventory.allinventoryrowswithnewlotshouldhaverealqtygreatterthanzero.msg                        = All inventory(ies) row(s) with new lot should have real quantity greater than zero before validation.
```
```suggestion
angal.inventory.allinventoryrowswithnewlotshouldhaverealqtygreatterthanzero.msg                        = All inventory(ies) row(s) with new lot should have a real quantity greater than zero before validation.
```

dbmalkovsky commented on bundle/language_en.properties:
```diff
@@ -667,6 +677,116 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
+angal.inventory.pleaseselectonlyoneinventory.msg                                                       = Please select only one inventory.
+angal.inventory.pleasevalidateinventoryagainsbeforeconfirmation.msg                                    = Please validate the inventory again before confirming it.
+angal.inventory.printing.error.msg                                                                     = An error occurred while printing.
+angal.inventory.productalreadyexist.fmt.msg                                                            = {0} already exist in the table. Do you want to add it anyway?
```
```suggestion
angal.inventory.productalreadyexist.fmt.msg                                                            = {0} already exists in the table. Do you want to add it anyway?
```

mwithi committed:
Apply suggestions from code review

Co-authored-by: David B Malkovsky <david@malkovsky.org>

mwithi committed:
Apply suggestions from code review

dbmalkovsky reviewed (approved):
LGTM
